python/k4FWCore/ApplicationMgr.py

```diff
@@ -37,6 +40,64 @@ class ApplicationMgr:
     def __init__(self, **kwargs):
         self._mgr = AppMgr(**kwargs)
 
+    def _setup_reader(self, reader, iosvc_props):
+        """Setup the reader consistently such that it has sane defaults
+
+        In some cases we have to peek into the files to obtain some information
+        to set sane default values if they are not set by the user. We need to
+        obtain
+        - The number of events in case EvtMax is set to -1
+        - The collection names in case the CollectionNames are not provided
+
+        Knowing the number of events is necessary to avoid errors when running
+        multithreaded since if we have, for example, 10 events and we are
+        running 9 at the same time, then (possibly) the first 9 complete and 9
+        more are scheduled, out of which only one will be finished without
+        errors. If we know the number of events in advance then we can just
+        schedule those.
+
+        We need the collection names to read to feed them to the Reader. Either
+        we take the user provided ones or we get them from the first event in
+        the file we peek into.
+        """
+        # First we determine whether we need to peek at all
+        inp = None
+        if iosvc_props["input"][0]:
+            inp = "input"
+        elif iosvc_props["Input"][0]:
+            inp = "Input"
+
+        if not inp:
+            # We have got nothing to do here, since there is no input
+            return
+
+        collections = iosvc_props["CollectionNames"][0] or None
+        n_events = self._mgr.EvtMax
+
+        if collections and n_events != -1:
+            # We know everything we need
+            logger.info(f"Initializing reader to read {n_events} events and read: {collections}")
+            reader.InputCollections = collections
+            return
```

Comment:
```suggestion
        if collections:
            reader.InputCollections = collections
            if n_events != -1:
                # We know everything we need
                logger.info(f"Initializing reader to read {n_events} events and read: {collections}")
                return
```
If any collections are passed then set the collections unconditionally, this is how it was being done before. I'll make a comment below.

Reply:
Oh yeah, good catch. I completely missed that. I have added a test that covers this scenario as well now. The failure I get without this fix is also what made me put in the "fixes" for the Writer that are not strictly necessary, because otherwise the writer would try to get non-existent collections.